### PR TITLE
Feature/check mandatory headers in hmac param

### PR DESCRIPTION
**Pull request by vaibhavatul47**
### Summary
Making date or x-date header mandatory in Hmac-headers.

#### Mandatory header values in Hmac-headers parameter

Currently, although the plugin validates clock-skew but it is no guarantee that the date or x-date header used in clock skew validation is not tampered.
Also, according to [ietf draft](https://tools.ietf.org/html/draft-cavage-http-signatures-05#section-2.3)  it is madatory to include date headers in Hmac-header parameter.
If `validate-request-body` config is enabled for an API then Hmac-header should also must contain digest header.
**Note:** `validate-request-body` config is not present in this code. I'll add it in another PR as it is associated with another feature.

### Test edited
* should not pass with GET with date header missing from hmac-headers

### CODE REVIEW

**Comment by p0pr0ck5** [comment]
Changes to the codebase should try to be as atomic as possible. Is it possible to re-work this PR such that it does not rely on a feature present in another branch?

**Comment by vaibhavatul47** [comment]
@p0pr0ck5 I can remove `validate-request-body` part from this PR. Then, once this PR is merged, I'll work on PR #2419 and make corresponding changes in that PR.
This will suffice your atomic condition for PR. #2418  will comprise of changes related to mandating hmac-headers and #2419 will comprise of request body validation changes.

**Comment by p0pr0ck5** [comment]
@vaibhavatul47 ok, sounds good. will you make these changes?

**Comment by vaibhavatul47** [comment]
@p0pr0ck5 sure, I'll update once its ready.

**Comment by shashiranjan84** [comment]
@vaibhavatul47 here client has option to include `Date` or `X-Date` header as part of signing string.  Here onus is on client to include `Date` or `X-Date` not Kong to mandate it.

File: kong/plugins/hmac-auth/access.lua [hunk 1]
```diff
@@ -13,21 +13,24 @@ local ngx_sha1 = ngx.hmac_sha1
 local ngx_set_header = ngx.req.set_header
 local ngx_get_headers = ngx.req.get_headers
 local ngx_log = ngx.log
+local req_read_body = ngx.req.read_body
+local req_get_body_data = ngx.req.get_body_data
```
Comment (thibaultcha) [inline kong/plugins/hmac-auth/access.lua]: @p0pr0ck5 @shashiranjan84 Will you make sure that this is using the recent public API draft to read the request body as well? Thanks!

**Comment by vaibhavatul47** [comment]
@shashiranjan84 Yes client would still have the option to chose from `Date` or `X-Date` header, but at least one of them must be used in hmac signing string generation.

**Comment by shashiranjan84** [comment]
@vaibhavatul47 no as per draft, it is must for client to send the `date` header 

"In order to generate the string that is signed with a key, the **client**
   MUST use the values of each HTTP header field in the `headers`
   Signature parameter, in the order they appear in the `headers`
   Signature parameter.  It is out of scope for this document to dictate
   what header fields an application will want to enforce, but
   implementers SHOULD at minimum include the request target and Date
   header fields."

**Comment by vaibhavatul47** [comment]
@shashiranjan84 I never meant that server should send `date` header. I'm trying to say, whatever request client is sending, it must contain a header with date information.

**Comment by shashiranjan84** [comment]
@vaibhavatul47 Sorry I meant, It on client to make sure to add `date` and `request-line` in order to make http-signature, not on Kong to force client to include these headers.

**Comment by shashiranjan84** [comment]
I would prefer updating doc, to inform user to make sure to include `date` and `request-line` in signing string, over forcing it from Kong side.

**Comment by vaibhavatul47** [comment]
@shashiranjan84 I think you are right, we should not mandate `date` header for following two reasons:
1. The change is not backwards compatible. If a user is currently using Kong without `date` header in hmac signature, then she/he would face validation error.
2. Not all users are concerned about replay attacks so it makes sense to leave it to users to include `date` header in hmac signature as per their requirement.

I will update my code to following:
1. `date` or `x-date` headers won't be mandatory.
2. If `validate-request-body` config is enabled, then only `digest` header will be mandated.

Also, we will need to update in docs to encourage users to include `date` header in the signature.

**Comment by shashiranjan84** [comment]
@vaibhavatul47 I am thinking to close this PR and update your another PR #2419 to make it work independent of this PR. I will take care of  the addition changes.

**Comment by shashiranjan84** [comment]
As per discussion before, I am closing this.

**Pull request closed without merging**